### DISCUSSION TITLE
[Feature]: 提供更加符合直觉和优雅的配置

### DISCUSSION THREAD

**Discussion by zhangfisher** [web-infra-dev/modern.js#4270] [General]
#### 多构建产物时的配置

目前当需要配置多个构建产物时是这样的:
```
export default defineConfig({
  plugins: [moduleTools(), ModulePlugin()],
  buildConfig: [
    {
      input: ['src'],
      format: 'esm',
      splitting: true,
      sourceMap: false,
      minify: 'esbuild',
      dts: false,
      target: 'es2021',
    },
    {
      input: ['src'],
      format: 'cjs',
      splitting: true,
      sourceMap: false,
      minify: 'esbuild',
      dts: false,
      target: 'es2021',
    },
  ]
}) 
```
显然，配置的冗余比较多。而当我们在构建时多个构建产物是基本需求。

#### 预设配置定制
 
目前预设配置定制采用`buildPreset`和`buildConfig`两个配置项，然后两者在冲突时进行警告提供的方式不够优雅
完全可以将两个配置合为一个

#### 期望:  多构建产物时的配置

```
export default defineConfig({
  plugins: [moduleTools(), ModulePlugin()],
  buildConfig: [
    {
      input: ['src'],
      format: ['esm','cjs','umd','iife'],
      splitting: true,
      sourceMap: false,
      minify: 'esbuild',
      dts: false,
      target: 'es2021',
    }
  ]
}) 
```
只需要指定一个`format`参数即可。

####  期望:预设配置定制

可以将`buildPreset`和`buildConfig`两个配置项合为一个

```
export default defineConfig({
  plugins: [moduleTools(), ModulePlugin()],
  buildConfig: [
    {
      preset:""npm-library"   //可选配置
     // 以下所有配置均可以覆盖预设配置 
     input: ['src'],
      format: ['esm','cjs','umd','iife'],
      splitting: true,
      sourceMap: false,
      minify:true,         // =true时采用默认压缩方式，如esbuild
      dts: false,
      target: 'es2021',
    },
  {
      preset:""npm-library"   //可选配置
      input: ['lib'],
    }
  ]
}) 
```
这种方法比起原来的方式更加简单优雅

像`sourceMap`和`minify`这种配置均应该支持`=true`的配置，采用推荐的最佳配置
比如`minify=true`时就默认采用`esbuild`，而不是一定要配一个`esbuild`或`tersert`,说实话，大部份情况下使用`esbuild`或`tersert`根本不重要，很多人甚至没听过`tersert`，只需要简单地配置`minify=true`就很香了

**Comment by chenjiahan** [comment]
感谢建议！

> 配置的冗余比较多。而当我们在构建时多个构建产物是基本需求

目前不同 format 构建产物的配置是分开的，这样的好处是可以对两种 format 的产物做灵活的配置，如果两者的大部分配置是相同的，那抽取一个 commonConfig 对象出来即可。

如果合并成 `format` 数组，那么就没法对两种 `format` 做不同的配置了。

> 可以将buildPreset和buildConfig两个配置项合为一个

我觉得这里主要考虑到 buildPreset 并不是一个对象，buildPreset 里面通常包含了多个 buildConfig，所以没法改造成你建议的形式。

> 像sourceMap和minify这种配置均应该支持=true的配置

赞同，`true` 确实更简单。

**Comment by chenjiahan** [comment]
@10Derozan 配置这块你有什么想法不

**Comment by justrustc** [comment]
改成这样的如何？
```
{
    buildConfig: {
    // 最外层的为公共配置
    input: ['src'],
    splitting: true,
    sourceMap: false,
    minify: true,
    dts: false,
    target: 'es2021',
    esm: {
      // 覆盖公共配置
      sourceMap: true,
    },
    cjs: {},
    umd: {},
  },
}
```

**Comment by 10Derozan** [comment]
> @10Derozan 配置这块你有什么想法不

我觉得将 `buildConfig` 变成可以覆盖 `buildPreset` 里的配置是相比目前处理「两者不能同时出现」更好的方式。
Preset 是对于一个「package」构建的预设，因此包含多份构建配置是我们预期的。
另外，至于`format`的形式，可以支持数组形式，目前我们推荐 ts 配置文件的形式，相同配置抽一下就好。

**Comment by 10Derozan** [comment]
> 改成这样的如何？
> 
> ```
> {
>     buildConfig: {
>     // 最外层的为公共配置
>     input: ['src'],
>     splitting: true,
>     sourceMap: false,
>     minify: true,
>     dts: false,
>     target: 'es2021',
>     esm: {
>       // 覆盖公共配置
>       sourceMap: true,
>     },
>     cjs: {},
>     umd: {},
>   },
> }
> ```

当然是ok的，同时这会加深对象的层级，以及强化了 `format` ，无法再提供一个默认值

**Comment by zhangfisher** [comment]
export default defineConfig({
  plugins: [moduleTools(), ModulePlugin()],
```
  buildConfig: [
    {
      preset:""npm-library"   //可选配置
     // 以下所有配置均可以覆盖预设配置 
     input: ['src'],
      format: [
          'esm',
          'cjs',
           'umd',                  //  简单的字符串
           {
                type:"cjs",
               sourceMap:false 
              //......
            }
     ],
      splitting: true,
      sourceMap: false,
      minify:true,         // =true时采用默认压缩方式，如esbuild
      dts: false,
      target: 'es2021',
    },
  {
      preset:""npm-library"   //可选配置
      input: ['lib'],
    }
  ]
}) 
```

**Comment by chenjiahan** [comment]
> 我觉得将 buildConfig 变成可以覆盖 buildPreset 里的配置是相比目前处理「两者不能同时出现」更好的方式。

是的，babel 的 config 就是可以覆盖 preset 的，两者会进行 merge
